org.spoofax.terms/src/org/spoofax/interpreter/terms/ISimpleTerm.java

```diff
@@ -37,26 +37,84 @@
 	 */
-    @Nullable <T extends ITermAttachment> T getAttachment(@Nullable TermAttachmentType<T> type);
+    @SuppressWarnings("unchecked")
+	default @Nullable <T extends ITermAttachment> T getAttachment(@Nullable TermAttachmentType<T> type) {
+		ITermAttachment attachment = internalGetAttachment();
+		if(type == null)
+			return (T) attachment;
+		for(ITermAttachment a = attachment; a != null; a = a.getNext()) {
+			if(a.getAttachmentType() == type)
+				return (T) a;
+		}
+		return null;
+	}
 
     // FIXME: Return a new term when adding/changing an attachment. Make the term immutable.
     /**
      * Adds the specified term attachment to this term.
      * If a term attachment of this type is already present, it is replaced.
      *
-     * @param resourceAttachment the attachment to add
+     * @param attachmentToPut the attachment to add
      */
-	void putAttachment(ITermAttachment resourceAttachment);
+	default void putAttachment(ITermAttachment attachmentToPut) {
+		ITermAttachment attachment = internalGetAttachment();
+		if(attachmentToPut == null)
+			return;
+		if(attachment == null) {
+			internalSetAttachment(attachmentToPut);
+		} else {
+			TermAttachmentType<?> newType = attachmentToPut.getAttachmentType();
+			if(attachment.getAttachmentType() == newType) {
+				attachmentToPut.setNext(attachment.getNext());
+				internalSetAttachment(attachmentToPut);
+			} else {
+				ITermAttachment previous = attachment;
+				for(ITermAttachment a = previous.getNext(); a != null; a = a.getNext()) {
+					if(a.getAttachmentType() == newType) {
+						attachmentToPut.setNext(a.getNext());
+						previous.setNext(attachmentToPut);
+						break;
+					}
+					previous = a;
+				}
+				previous.setNext(attachmentToPut);
+			}
+		}
+	}
 
     // FIXME: Return a new term when removing an attachment. Make the term immutable.
     /**
      * Removes the term attachment of the specified type,
      * if one is available for this term.
      *
-     * @param attachmentType the term attachment type
+     * @param type the term attachment type
      * @return the removed term attachment; or {@code null} when no attachment was removed
      */
-    @Nullable ITermAttachment removeAttachment(TermAttachmentType<?> attachmentType);
+    default @Nullable ITermAttachment removeAttachment(TermAttachmentType<?> type) {
+		ITermAttachment attachment = internalGetAttachment();
+		if(attachment != null) {
+			if(attachment.getAttachmentType() == type) {
+				internalSetAttachment(attachment.getNext());
+				attachment.setNext(null);
+				return attachment;
+			} else {
+				ITermAttachment previous = attachment;
+				for(ITermAttachment a = attachment.getNext(); a != null; a = a.getNext()) {
+					if(a.getAttachmentType() == type) {
+						previous.setNext(a.getNext());
+						a.setNext(null);
+						return a;
+					}
+					previous = a;
+				}
+			}
+		}
+		return null;
+	}
+
+    ITermAttachment internalGetAttachment();
```

Comment:
I don't think exposing internals on ISimpleTerm is a good idea, since just about everything inherits from this? Why is a change to ISimpleTerm needed, I thought this was about IStrategoArrayList?

Reply:
This allowed me to pull up the implementations of `getAttachment` and `putAttachment` from `AbstractSimpleTerm`, so that code can be shared by other implementations of `ISimpleTerm` without having to extend `AbstractSimpleTerm`. In particular, `StrategoArrayListTail` does not have a field with an `attachment` like `AbstractSimpleTerm` does (memory optimisation - the attachments live in a field of the original `StrategoArrayList` that the tail object points to) but does need the same logic for get/put on attachments.